File: python/context_engineering/webhook.py
"""
Webhook Telemetry Integration

Fire-and-forget webhook reporting for pack/trace/handoff/pipeline/quality/cost.
Designed for Make.com scenario ingestion but works with any HTTP endpoint.

Usage::

    reporter = create_webhook_reporter(analytics_url="https://...")
    result = pack(items, budget)
    reporter.report_pack(result)

Or rely on env vars::

    # reads CE_WEBHOOK_URL, CE_WEBHOOK_HANDOFF_URL, CE_WEBHOOK_QUALITY_URL, CE_WEBHOOK_COST_URL
    reporter = create_webhook_reporter()
"""

from __future__ import annotations

import json
import logging
import os
import threading
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional

import httpx

from .core import ContextPack, ContextTrace
from .cost import CostEstimate
from .quality import ContextQuality

logger = logging.getLogger(__name__)

# ─── Payload Types ────────────────────────────────────────────────────


@dataclass
class WebhookAnalyticsPayload:
    event_type: str  # "pack" | "trace"
    session_id: str
    model: str
    strategy: str
    timestamp: str
    budget_max_tokens: int
    budget_reserve_tokens: int
    total_tokens: int
    selected_count: int
    dropped_count: int
    budget_utilization_pct: float
    remaining_tokens: int
    quality_overall: Optional[float] = None
    quality_density: Optional[float] = None
    quality_diversity: Optional[float] = None
    cost_with_cache: Optional[float] = None
    cost_without_cache: Optional[float] = None
    cache_hit_ratio: Optional[float] = None
    trace_decisions: Optional[str] = None


@dataclass
class WebhookHandoffPayload:
    event_type: str  # "handoff"
    session_id: str
    timestamp: str
    total_issues: int
    active_items: int
    deferred_items: int
    source_agent: Optional[str] = None
    target_url: Optional[str] = None
    jsonl_size_bytes: int = 0


@dataclass
class WebhookPipelinePayload:
    event_type: str  # "pipeline"
    session_id: str
    timestamp: str
    model: str
    strategy: str
    stages: List[str] = field(default_factory=list)
    input_count: int = 0
    selected_count: int = 0
    dropped_count: int = 0
    total_tokens: int = 0
    budget_max_tokens: int = 0
    budget_reserve_tokens: int = 0
    budget_utilization_pct: float = 0
    quality_overall: Optional[float] = None
    quality_density: Optional[float] = None
    quality_diversity: Optional[float] = None
    cache_efficiency: Optional[float] = None
    cacheable_tokens: Optional[int] = None
    cache_key: Optional[str] = None
    allocation_efficiency: Optional[float] = None
    allocations_json: Optional[str] = None
    placement_strategy: Optional[str] = None


@dataclass
class WebhookQualityPayload:
    event_type: str  # "quality"
    session_id: str
    timestamp: str
    model: str
    quality_overall: float = 0
    quality_density: float = 0
    quality_diversity: float = 0
    selected_count: int = 0
    dropped_count: int = 0
    budget_utilization_pct: float = 0


@dataclass
class WebhookCostPayload:
    event_type: str  # "cost"
    session_id: str
    timestamp: str
    model: str
    cost_with_cache: float = 0
    cost_without_cache: float = 0
    cache_hit_ratio: float = 0
    total_tokens: int = 0
    budget_max_tokens: int = 0


@dataclass
class PackReportExtras:
    quality: Optional[ContextQuality] = None
    cost: Optional[CostEstimate] = None
    cache_hit_ratio: Optional[float] = None


@dataclass
class HandoffReportExtras:
    source_agent: Optional[str] = None
    target_url: Optional[str] = None


@dataclass
class PipelineReportExtras:
    cost: Optional[CostEstimate] = None
    cache_hit_ratio: Optional[float] = None
    placement_strategy: Optional[str] = None


@dataclass
class WebhookOptions:
    analytics_url: Optional[str] = None
    handoff_url: Optional[str] = None
    quality_url: Optional[str] = None
    cost_url: Optional[str] = None
    session_id: Optional[str] = None
    model: Optional[str] = None
    strategy: Optional[str] = None
    timeout_s: float = 5.0
    headers: Dict[str, str] = field(default_factory=dict)


# ─── Internal Helpers ─────────────────────────────────────────────────


def _compute_utilization(total_tokens: int, max_tokens: int, reserve_tokens: int) -> float:
    effective = max_tokens - reserve_tokens
    return round((total_tokens / effective) * 10000) / 100 if effective > 0 else 0


def _fire_webhook(
    url: str,
    payload: Dict[str, Any],
    timeout_s: float,
    headers: Dict[str, str],
) -> None:
    """Send webhook in a daemon thread (fire-and-forget)."""

    def _send() -> None:
        try:
            httpx.post(
                url,
                json=payload,
                headers={"Content-Type": "application/json", **headers},
                timeout=timeout_s,
            )
        except Exception as exc:
            logger.warning("webhook:send_failed url=%s error=%s", url, exc)

    t = threading.Thread(target=_send, daemon=True)
    t.start()


def _strip_none(d: Dict[str, Any]) -> Dict[str, Any]:
    """Remove None values from a dict for cleaner payloads."""
    return {k: v for k, v in d.items() if v is not None}


def _build_analytics_payload(
    pack: ContextPack,
    event_type: str,
    session_id: str,
    model: str,
    strategy: str,
    extras: Optional[PackReportExtras] = None,
    trace_decisions: Optional[str] = None,
) -> Dict[str, Any]:
    max_tokens = pack.budget.max_tokens
    reserve_tokens = pack.budget.reserve_tokens or 0
    utilization_pct = _compute_utilization(pack.total_tokens, max_tokens, reserve_tokens)

    payload: Dict[str, Any] = {
        "event_type": event_type,
        "session_id": session_id,
        "model": model,
        "strategy": strategy,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "budget_max_tokens": max_tokens,
        "budget_reserve_tokens": reserve_tokens,
        "total_tokens": pack.total_tokens,
        "selected_count": len(pack.selected),
        "dropped_count": len(pack.dropped),
        "budget_utilization_pct": utilization_pct,
        "remaining_tokens": max(0, max_tokens - reserve_tokens - pack.total_tokens),
    }

    if extras:
        if extras.quality:
            payload["quality_overall"] = extras.quality.overall
            payload["quality_density"] = extras.quality.density
            payload["quality_diversity"] = extras.quality.diversity
        if extras.cost:
            payload["cost_with_cache"] = extras.cost.cost_with_cache
            payload["cost_without_cache"] = extras.cost.cost_without_cache
        if extras.cache_hit_ratio is not None:
            payload["cache_hit_ratio"] = extras.cache_hit_ratio

    if trace_decisions is not None:
        payload["trace_decisions"] = trace_decisions

    return payload


def _build_pipeline_payload(
    result: Any,
    session_id: str,
    model: str,
    strategy: str,
    extras: Optional[PipelineReportExtras] = None,
) -> Dict[str, Any]:
    max_tokens = result.budget.max_tokens
    reserve_tokens = result.budget.reserve_tokens or 0

    payload: Dict[str, Any] = {
        "event_type": "pipeline",
        "session_id": session_id,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "model": model,
        "strategy": strategy,
        "stages": list(result.stages),
        "input_count": result.input_count,
        "selected_count": len(result.selected),
        "dropped_count": len(result.dropped),
        "total_tokens": result.total_tokens,
        "budget_max_tokens": max_tokens,
        "budget_reserve_tokens": reserve_tokens,
        "budget_utilization_pct": _compute_utilization(
            result.total_tokens, max_tokens, reserve_tokens
        ),
    }

    if result.quality is not None:
        payload["quality_overall"] = result.quality.overall
        payload["quality_density"] = result.quality.density
        payload["quality_diversity"] = result.quality.diversity
    if result.cache_efficiency is not None:
        payload["cache_efficiency"] = result.cache_efficiency
    if result.cacheable_tokens is not None:
        payload["cacheable_tokens"] = result.cacheable_tokens
    if result.cache_key is not None:
        payload["cache_key"] = result.cache_key
    if result.allocation_efficiency is not None:
        payload["allocation_efficiency"] = result.allocation_efficiency
    if result.allocations is not None:
        payload["allocations_json"] = json.dumps(result.allocations)
    if extras and extras.placement_strategy:
        payload["placement_strategy"] = extras.placement_strategy

    return payload


# ─── Reporter Protocol ────────────────────────────────────────────────


class WebhookReporter:
    """Webhook reporter that sends telemetry for pack/trace/handoff/pipeline/quality/cost."""

    def __init__(
        self,
        analytics_url: str,
        handoff_url: str,
        quality_url: str,
        cost_url: str,
        session_id: str,
        model: str,
        strategy: str,
        timeout_s: float,
        headers: Dict[str, str],
    ) -> None:
        self._analytics_url = analytics_url
        self._handoff_url = handoff_url
        self._quality_url = quality_url
        self._cost_url = cost_url
        self._session_id = session_id
        self._model = model
        self._strategy = strategy
        self._timeout_s = timeout_s
        self._headers = headers

    def report_pack(
        self,
        pack: ContextPack,
        extras: Optional[PackReportExtras] = None,
    ) -> None:
        if not self._analytics_url:
            return
        payload = _build_analytics_payload(
            pack, "pack", self._session_id, self._model, self._strategy, extras
        )
        _fire_webhook(self._analytics_url, payload, self._timeout_s, self._headers)

    def report_trace(
        self,
        trace: ContextTrace,
        extras: Optional[PackReportExtras] = None,
    ) -> None:
        if not self._analytics_url:
            return
        decisions = json.dumps(
            [
                {
                    "id": s.id,
                    "decision": s.decision,
                    "tokens": s.tokens,
                    "reason": s.reason,
                }
                for s in trace.steps
            ]
        )
        payload = _build_analytics_payload(
            trace.pack,
            "trace",
            self._session_id,
            self._model,
            self._strategy,
            extras,
            decisions,
        )
        _fire_webhook(self._analytics_url, payload, self._timeout_s, self._headers)

    def report_handoff(
        self,
        handoff: Any,
        extras: Optional[HandoffReportExtras] = None,
    ) -> None:
        if not self._handoff_url:
            return
        stats = handoff.stats if hasattr(handoff, "stats") else handoff.get("stats", {})
        jsonl = handoff.jsonl if hasattr(handoff, "jsonl") else handoff.get("jsonl", "")
        payload: Dict[str, Any] = {
            "event_type": "handoff",
            "session_id": self._session_id,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "total_issues": stats.get("totalIssues", 0)
            if isinstance(stats, dict)
            else getattr(stats, "total_issues", 0),
            "active_items": stats.get("activeItems", 0)
            if isinstance(stats, dict)
            else getattr(stats, "active_items", 0),
            "deferred_items": stats.get("deferredItems", 0)
            if isinstance(stats, dict)
            else getattr(stats, "deferred_items", 0),
            "jsonl_size_bytes": len(jsonl.encode("utf-8")),
        }
        if extras:
            if extras.source_agent:
                payload["source_agent"] = extras.source_agent
            if extras.target_url:
                payload["target_url"] = extras.target_url
        _fire_webhook(self._handoff_url, payload, self._timeout_s, self._headers)

    def report_pipeline(
        self,
        result: Any,
        extras: Optional[PipelineReportExtras] = None,
    ) -> None:
        if not self._analytics_url:
            return
        payload = _build_pipeline_payload(
            result, self._session_id, self._model, self._strategy, extras
        )
        _fire_webhook(self._analytics_url, payload, self._timeout_s, self._headers)

    def report_quality(
        self,
        pack: ContextPack,
        quality: ContextQuality,
    ) -> None:
        if not self._quality_url:
            return
        max_tokens = pack.budget.max_tokens
        reserve_tokens = pack.budget.reserve_tokens or 0
        payload: Dict[str, Any] = {
            "event_type": "quality",
            "session_id": self._session_id,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "model": self._model,
            "quality_overall": quality.overall,
            "quality_density": quality.density,
            "quality_diversity": quality.diversity,
            "selected_count": len(pack.selected),
            "dropped_count": len(pack.dropped),
            "budget_utilization_pct": _compute_utilization(
                pack.total_tokens, max_tokens, reserve_tokens
            ),
        }
        _fire_webhook(self._quality_url, payload, self._timeout_s, self._headers)

    def report_cost(
        self,
        pack: ContextPack,
        cost: CostEstimate,
        cache_hit_ratio: Optional[float] = None,
    ) -> None:
        if not self._cost_url:
            return
        payload: Dict[str, Any] = {
            "event_type": "cost",
            "session_id": self._session_id,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "model": self._model,
            "cost_with_cache": cost.cost_with_cache,
            "cost_without_cache": cost.cost_without_cache,
            "cache_hit_ratio": cache_hit_ratio
            if cache_hit_ratio is not None
            else cost.cache_efficiency,
            "total_tokens": cost.input_tokens,
            "budget_max_tokens": pack.budget.max_tokens,
        }
        _fire_webhook(self._cost_url, payload, self._timeout_s, self._headers)


class _NoopReporter:
    """No-op reporter that never sends anything."""

    def report_pack(self, pack: Any, extras: Any = None) -> None:
        pass

    def report_trace(self, trace: Any, extras: Any = None) -> None:
        pass

    def report_handoff(self, handoff: Any, extras: Any = None) -> None:
        pass

    def report_pipeline(self, result: Any, extras: Any = None) -> None:
        pass

    def report_quality(self, pack: Any, quality: Any) -> None:
        pass

    def report_cost(self, pack: Any, cost: Any, cache_hit_ratio: Any = None) -> None:
        pass


noop_reporter = _NoopReporter()


# ─── Factory ──────────────────────────────────────────────────────────


def create_webhook_reporter(
    analytics_url: Optional[str] = None,
    handoff_url: Optional[str] = None,
    quality_url: Optional[str] = None,
    cost_url: Optional[str] = None,
    session_id: Optional[str] = None,
    model: Optional[str] = None,
    strategy: Optional[str] = None,
    timeout_s: float = 5.0,
    headers: Optional[Dict[str, str]] = None,
) -> WebhookReporter | _NoopReporter:
    """Create a webhook reporter for telemetry.

    Reads env vars as defaults:
    - ``CE_WEBHOOK_URL`` -- pack/trace analytics
    - ``CE_WEBHOOK_HANDOFF_URL`` -- handoff events
    - ``CE_WEBHOOK_QUALITY_URL`` -- quality regression reports
    - ``CE_WEBHOOK_COST_URL`` -- cost anomaly reports

    Returns a no-op reporter when no URL is configured.
    """
    resolved_analytics = analytics_url or os.environ.get("CE_WEBHOOK_URL", "")
    resolved_handoff = handoff_url or os.environ.get("CE_WEBHOOK_HANDOFF_URL", "")
    resolved_quality = quality_url or os.environ.get("CE_WEBHOOK_QUALITY_URL", "")
    resolved_cost = cost_url or os.environ.get("CE_WEBHOOK_COST_URL", "")

    if (
        not resolved_analytics
        and not resolved_handoff
        and not resolved_quality
        and not resolved_cost
    ):
        return noop_reporter

    import time

    resolved_session = session_id or f"ce-{int(time.time()):x}"

    return WebhookReporter(
        analytics_url=resolved_analytics,
        handoff_url=resolved_handoff,
        quality_url=resolved_quality,
        cost_url=resolved_cost,
        session_id=resolved_session,
        model=model or "unknown",
        strategy=strategy or "greedy-score",
        timeout_s=timeout_s,
        headers=headers or {},
    )

File: packages/ce-core/src/webhook.ts
/**
 * Webhook Telemetry Integration
 *
 * Fire-and-forget webhook reporting for pack/trace/handoff/pipeline/quality/cost.
 * Designed for Make.com scenario ingestion but works with any HTTP endpoint.
 *
 * Usage:
 * ```ts
 * const reporter = createWebhookReporter({ analyticsUrl: "https://..." });
 * const result = pack(items, budget);
 * reporter.reportPack(result);
 * ```
 *
 * Or rely on env vars:
 * ```ts
 * // reads CE_WEBHOOK_URL, CE_WEBHOOK_HANDOFF_URL, CE_WEBHOOK_QUALITY_URL, CE_WEBHOOK_COST_URL
 * const reporter = createWebhookReporter();
 * ```
 */

import type { ContextPack, ContextTrace } from "./types.js";
import type { ContextQuality } from "./quality.js";
import type { CostEstimate } from "./cost.js";
import type { HandoffResult } from "./beads.js";
import type { PipelineResult } from "./pipeline.js";
import type { Logger } from "./logger.js";
import { noopLogger } from "./logger.js";

// ─── Types ───────────────────────────────────────────────────────────

export interface WebhookOptions {
  /** URL for analytics/telemetry webhook (pack/trace results) */
  analyticsUrl?: string;
  /** URL for handoff webhook (BEADS handoff notifications) */
  handoffUrl?: string;
  /** URL for quality regression webhook */
  qualityUrl?: string;
  /** URL for cost anomaly webhook */
  costUrl?: string;
  /** Session identifier for correlation */
  sessionId?: string;
  /** Model name for telemetry */
  model?: string;
  /** Packing strategy label */
  strategy?: string;
  /** Logger for warning on failures */
  logger?: Logger;
  /** Fetch timeout in milliseconds (default: 5000) */
  timeoutMs?: number;
  /** Additional headers to send with requests */
  headers?: Record<string, string>;
}

export interface WebhookAnalyticsPayload {
  event_type: "pack" | "trace";
  session_id: string;
  model: string;
  strategy: string;
  timestamp: string;
  budget_max_tokens: number;
  budget_reserve_tokens: number;
  total_tokens: number;
  selected_count: number;
  dropped_count: number;
  budget_utilization_pct: number;
  remaining_tokens: number;
  quality_overall?: number;
  quality_density?: number;
  quality_diversity?: number;
  cost_with_cache?: number;
  cost_without_cache?: number;
  cache_hit_ratio?: number;
  trace_decisions?: string;
}

export interface WebhookHandoffPayload {
  event_type: "handoff";
  session_id: string;
  timestamp: string;
  total_issues: number;
  active_items: number;
  deferred_items: number;
  source_agent?: string;
  target_url?: string;
  jsonl_size_bytes: number;
}

export interface WebhookPipelinePayload {
  event_type: "pipeline";
  session_id: string;
  timestamp: string;
  model: string;
  strategy: string;
  stages: string[];
  input_count: number;
  selected_count: number;
  dropped_count: number;
  total_tokens: number;
  budget_max_tokens: number;
  budget_reserve_tokens: number;
  budget_utilization_pct: number;
  quality_overall?: number;
  quality_density?: number;
  quality_diversity?: number;
  cache_efficiency?: number;
  cacheable_tokens?: number;
  cache_key?: string;
  allocation_efficiency?: number;
  allocations_json?: string;
  placement_strategy?: string;
}

export interface WebhookQualityPayload {
  event_type: "quality";
  session_id: string;
  timestamp: string;
  model: string;
  quality_overall: number;
  quality_density: number;
  quality_diversity: number;
  selected_count: number;
  dropped_count: number;
  budget_utilization_pct: number;
}

export interface WebhookCostPayload {
  event_type: "cost";
  session_id: string;
  timestamp: string;
  model: string;
  cost_with_cache: number;
  cost_without_cache: number;
  cache_hit_ratio: number;
  total_tokens: number;
  budget_max_tokens: number;
}

export interface PackReportExtras {
  quality?: ContextQuality;
  cost?: CostEstimate;
  cacheHitRatio?: number;
}

export interface HandoffReportExtras {
  sourceAgent?: string;
  targetUrl?: string;
}

export interface PipelineReportExtras {
  cost?: CostEstimate;
  cacheHitRatio?: number;
  placementStrategy?: string;
}

export interface WebhookReporter {
  reportPack(pack: ContextPack, extras?: PackReportExtras): void;
  reportTrace(trace: ContextTrace, extras?: PackReportExtras): void;
  reportHandoff(handoff: HandoffResult, extras?: HandoffReportExtras): void;
  reportPipeline(result: PipelineResult, extras?: PipelineReportExtras): void;
  reportQuality(pack: ContextPack, quality: ContextQuality): void;
  reportCost(
    pack: ContextPack,
    cost: CostEstimate,
    cacheHitRatio?: number
  ): void;
}

// ─── No-op Reporter ──────────────────────────────────────────────────

/** Explicit no-op reporter that never sends anything. */
export const noopReporter: WebhookReporter = {
  reportPack() {},
  reportTrace() {},
  reportHandoff() {},
  reportPipeline() {},
  reportQuality() {},
  reportCost() {},
};

// ─── Internal Helpers ────────────────────────────────────────────────

function computeUtilization(
  totalTokens: number,
  maxTokens: number,
  reserveTokens: number
): number {
  const effective = maxTokens - reserveTokens;
  return effective > 0
    ? Math.round((totalTokens / effective) * 10000) / 100
    : 0;
}

function fireWebhook(
  url: string,
  payload: unknown,
  logger: Logger,
  timeoutMs: number,
  headers?: Record<string, string>
): void {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);

  fetch(url, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...headers,
    },
    body: JSON.stringify(payload),
    signal: controller.signal,
  })
    .then(() => {
      clearTimeout(timer);
    })
    .catch((err: unknown) => {
      clearTimeout(timer);
      const msg = err instanceof Error ? err.message : String(err);
      logger.warn("webhook:send_failed", { url, error: msg });
    });
}

function buildAnalyticsPayload(
  pack: ContextPack,
  eventType: "pack" | "trace",
  sessionId: string,
  model: string,
  strategy: string,
  extras?: PackReportExtras,
  traceDecisions?: string
): WebhookAnalyticsPayload {
  const maxTokens = pack.budget.maxTokens;
  const reserveTokens = pack.budget.reserveTokens ?? 0;
  const utilizationPct = computeUtilization(
    pack.totalTokens,
    maxTokens,
    reserveTokens
  );

  return {
    event_type: eventType,
    session_id: sessionId,
    model,
    strategy,
    timestamp: new Date().toISOString(),
    budget_max_tokens: maxTokens,
    budget_reserve_tokens: reserveTokens,
    total_tokens: pack.totalTokens,
    selected_count: pack.selected.length,
    dropped_count: pack.dropped.length,
    budget_utilization_pct: utilizationPct,
    remaining_tokens: Math.max(0, maxTokens - reserveTokens - pack.totalTokens),
    quality_overall: extras?.quality?.overall,
    quality_density: extras?.quality?.density,
    quality_diversity: extras?.quality?.diversity,
    cost_with_cache: extras?.cost?.costWithCache,
    cost_without_cache: extras?.cost?.costWithoutCache,
    cache_hit_ratio: extras?.cacheHitRatio,
    trace_decisions: traceDecisions,
  };
}

function buildPipelinePayload(
  result: PipelineResult,
  sessionId: string,
  model: string,
  strategy: string,
  extras?: PipelineReportExtras
): WebhookPipelinePayload {
  const maxTokens = result.budget.maxTokens;
  const reserveTokens = result.budget.reserveTokens ?? 0;

  const payload: WebhookPipelinePayload = {
    event_type: "pipeline",
    session_id: sessionId,
    timestamp: new Date().toISOString(),
    model,
    strategy,
    stages: result.stages,
    input_count: result.inputCount,
    selected_count: result.selected.length,
    dropped_count: result.dropped.length,
    total_tokens: result.totalTokens,
    budget_max_tokens: maxTokens,
    budget_reserve_tokens: reserveTokens,
    budget_utilization_pct: computeUtilization(
      result.totalTokens,
      maxTokens,
      reserveTokens
    ),
  };

  if (result.quality) {
    payload.quality_overall = result.quality.overall;
    payload.quality_density = result.quality.density;
    payload.quality_diversity = result.quality.diversity;
  }
  if (result.cacheEfficiency !== undefined) {
    payload.cache_efficiency = result.cacheEfficiency;
  }
  if (result.cacheableTokens !== undefined) {
    payload.cacheable_tokens = result.cacheableTokens;
  }
  if (result.cacheKey !== undefined) {
    payload.cache_key = result.cacheKey;
  }
  if (result.allocationEfficiency !== undefined) {
    payload.allocation_efficiency = result.allocationEfficiency;
  }
  if (result.allocations !== undefined) {
    payload.allocations_json = JSON.stringify(result.allocations);
  }
  if (extras?.placementStrategy) {
    payload.placement_strategy = extras.placementStrategy;
  }

  return payload;
}

// ─── Factory ─────────────────────────────────────────────────────────

/**
 * Create a webhook reporter for telemetry.
 *
 * Reads env vars as defaults:
 * - `CE_WEBHOOK_URL` — pack/trace analytics
 * - `CE_WEBHOOK_HANDOFF_URL` — handoff events
 * - `CE_WEBHOOK_QUALITY_URL` — quality regression reports
 * - `CE_WEBHOOK_COST_URL` — cost anomaly reports
 *
 * Returns no-op methods when no URL is configured.
 */
export function createWebhookReporter(
  options: WebhookOptions = {}
): WebhookReporter {
  const analyticsUrl = options.analyticsUrl ?? process.env.CE_WEBHOOK_URL ?? "";
  const handoffUrl =
    options.handoffUrl ?? process.env.CE_WEBHOOK_HANDOFF_URL ?? "";
  const qualityUrl =
    options.qualityUrl ?? process.env.CE_WEBHOOK_QUALITY_URL ?? "";
  const costUrl = options.costUrl ?? process.env.CE_WEBHOOK_COST_URL ?? "";
  const sessionId = options.sessionId ?? `ce-${Date.now().toString(36)}`;
  const model = options.model ?? "unknown";
  const strategy = options.strategy ?? "greedy-score";
  const logger = options.logger ?? noopLogger;
  const timeoutMs = options.timeoutMs ?? 5000;
  const headers = options.headers;

  if (!analyticsUrl && !handoffUrl && !qualityUrl && !costUrl) {
    return noopReporter;
  }

  return {
    reportPack(pack: ContextPack, extras?: PackReportExtras): void {
      if (!analyticsUrl) return;
      const payload = buildAnalyticsPayload(
        pack,
        "pack",
        sessionId,
        model,
        strategy,
        extras
      );
      fireWebhook(analyticsUrl, payload, logger, timeoutMs, headers);
    },

    reportTrace(trace: ContextTrace, extras?: PackReportExtras): void {
      if (!analyticsUrl) return;
      const traceDecisions = JSON.stringify(
        trace.steps.map(s => ({
          id: s.id,
          decision: s.decision,
          tokens: s.tokens,
          reason: s.reason,
        }))
      );
      const payload = buildAnalyticsPayload(
        trace.pack,
        "trace",
        sessionId,
        model,
        strategy,
        extras,
        traceDecisions
      );
      fireWebhook(analyticsUrl, payload, logger, timeoutMs, headers);
    },

    reportHandoff(handoff: HandoffResult, extras?: HandoffReportExtras): void {
      if (!handoffUrl) return;
      const payload: WebhookHandoffPayload = {
        event_type: "handoff",
        session_id: sessionId,
        timestamp: new Date().toISOString(),
        total_issues: handoff.stats.totalIssues,
        active_items: handoff.stats.activeItems,
        deferred_items: handoff.stats.deferredItems,
        source_agent: extras?.sourceAgent,
        target_url: extras?.targetUrl,
        jsonl_size_bytes: new TextEncoder().encode(handoff.jsonl).byteLength,
      };
      fireWebhook(handoffUrl, payload, logger, timeoutMs, headers);
    },

    reportPipeline(
      result: PipelineResult,
      extras?: PipelineReportExtras
    ): void {
      if (!analyticsUrl) return;
      const payload = buildPipelinePayload(
        result,
        sessionId,
        model,
        strategy,
        extras
      );
      fireWebhook(analyticsUrl, payload, logger, timeoutMs, headers);
    },

    reportQuality(pack: ContextPack, quality: ContextQuality): void {
      if (!qualityUrl) return;
      const maxTokens = pack.budget.maxTokens;
      const reserveTokens = pack.budget.reserveTokens ?? 0;
      const payload: WebhookQualityPayload = {
        event_type: "quality",
        session_id: sessionId,
        timestamp: new Date().toISOString(),
        model,
        quality_overall: quality.overall,
        quality_density: quality.density,
        quality_diversity: quality.diversity,
        selected_count: pack.selected.length,
        dropped_count: pack.dropped.length,
        budget_utilization_pct: computeUtilization(
          pack.totalTokens,
          maxTokens,
          reserveTokens
        ),
      };
      fireWebhook(qualityUrl, payload, logger, timeoutMs, headers);
    },

    reportCost(
      pack: ContextPack,
      cost: CostEstimate,
      cacheHitRatio?: number
    ): void {
      if (!costUrl) return;
      const payload: WebhookCostPayload = {
        event_type: "cost",
        session_id: sessionId,
        timestamp: new Date().toISOString(),
        model,
        cost_with_cache: cost.costWithCache,
        cost_without_cache: cost.costWithoutCache,
        cache_hit_ratio: cacheHitRatio ?? cost.cacheEfficiency,
        total_tokens: cost.inputTokens,
        budget_max_tokens: pack.budget.maxTokens,
      };
      fireWebhook(costUrl, payload, logger, timeoutMs, headers);
    },
  };
}
